src/main/java/io/cdap/cdap/metrics/process/gcp/GoogleCloudMonitoringWriter.java

```diff
@@ -0,0 +1,219 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metrics.process.gcp;
+
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.monitoring.v3.MetricServiceSettings;
+import com.google.gson.Gson;
+import com.google.monitoring.v3.TimeSeries;
+import io.cdap.cdap.api.metrics.MetricType;
+import io.cdap.cdap.api.metrics.MetricValue;
+import io.cdap.cdap.api.metrics.MetricValues;
+import io.cdap.cdap.api.metrics.MetricsWriter;
+import io.cdap.cdap.api.metrics.MetricsWriterContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileReader;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * {@link MetricsWriter} implementation for writing metrics to Google Cloud Monitoring
+ */
+public class GoogleCloudMonitoringWriter implements MetricsWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleCloudMonitoringWriter.class);
+  private static final Gson GSON = new Gson();
+  private static final String RESOURCE_CONTAINER = "resource_container";
+  private static final String PROJECT = "project";
+  private static final String ORG_ID = "org_id";
+  private static final String LOCATION = "location";
+  private static final String ANTHOS_CLUSTER = "anthos_cluster";
+  private static final String INSTANCE_ID = "instance_id";
+  private static final String VERSION = "version";
+  private static final String POLL_FREQUENCY_SECONDS = "poll.frequency.seconds";
+  private static final String MONITORING_ENDPOINT = "monitoring.endpoint";
+  private static final String CONFIG_FILE_PATH = "config.file.path";
+  private static final String WRITER_NAME = "google_cloud_monitoring_writer";
+
+  private final AtomicLong lastEndTime = new AtomicLong(-1);
+  private Map<String, MetricsMapping> metricsMapping;
+  @Nullable
+  private MetricServiceClient metricServiceClient;
+  @Nullable
+  private String projectName;
+  private Map<String, String> autoFilledLabelMap;
+  private int pollFreqInSeconds;
+
+  public GoogleCloudMonitoringWriter() {
+
+  }
+
+  @Override
+  public void write(Collection<MetricValues> metricValues) {
+    if (metricServiceClient == null) {
```

Comment:
nit: move these checks into a isInitialized() method? What you have is fine too, wrap these in a function seems to code more readable. Leave to you to decide.

Reply:
Done.